### PR TITLE
[release/2.1] Fix deadlock when waiting for process exit in Console.CancelKeyPress

### PR DESCRIPTION
**Pull request by stephentoub**
Port https://github.com/dotnet/corefx/pull/29709 to release/2.1
cc: @tmds, @danmosemsft 

Closes https://github.com/dotnet/corefx/issues/29699
NO MERGE until branch opened for 2.1.1.

### CODE REVIEW

**Comment by stephentoub** [comment]
@tmds, this change is failing in the release branch with native compilation errors.  Could you take a look?

File: src/Native/Unix/System.Native/pal_signal.h [hunk 1]
```diff
@@ -51,4 +51,12 @@ typedef void (*SigChldCallback)(int reapAll);
  */
 uint32_t SystemNative_RegisterForSigChld(SigChldCallback callback);
 
+/**
+ * Remove our handler and reissue the signal to be picked up by the previously registered handler.
+ *
+ * In the most common case, this will be the default handler, causing the process to be torn down.
+ * It could also be a custom handler registered by other code before us.
+ */
+DLLEXPORT void SystemNative_RestoreAndHandleCtrl(enum CtrlCode ctrlCode);
```
Comment (tmds) [inline src/Native/Unix/System.Native/pal_signal.h]: > 0:1: error: unknown type name 'DLLEXPORT'

DLLEXPORT was added post-2.1. Removing it should make the native build work.

File: src/Native/Unix/System.Native/pal_signal.c [hunk 2]
```diff
@@ -112,16 +112,14 @@ void* SignalHandlerLoop(void* arg)
         {
             // We're now handling SIGQUIT and SIGINT. Invoke the callback, if we have one.
             CtrlCallback callback = g_ctrlCallback;
-            int rv = callback != NULL ? callback(signalCode == SIGQUIT ? Break : Interrupt) : 0;
-            if (rv == 0) // callback removed or was invoked and didn't handle the signal
+            enum CtrlCode ctrlCode = signalCode == SIGQUIT ? Break : Interrupt;
```
Comment (danmosemsft) [inline src/Native/Unix/System.Native/pal_signal.c]: I'm not famliiar with this syntax (or I'm spacing). What does the enum keyword do here? Forward declaration?

Comment (tmds) [reply src/Native/Unix/System.Native/pal_signal.c]: It's like `struct` in C, if you don't add a `typedef` then you need to include the keyword also when declaring a variable.

**Comment by danmosemsft** [comment]
@dotnet-bot test Windows x64 Debug Build  please

**Comment by danmosemsft** [comment]
Branch opened; failures are unrelated known issues

**Pull request merged**